// angular-morris-js/esm2015/angular-morris-js.js
import { Directive, Input, ElementRef, NgZone, NgModule } from '@angular/core';

class MorrisChartDirective {
    constructor(elementRef, ngZone) {
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.window = window;
        this.type = 'Line';
    }
    ngOnInit() {
        this._options = Object.assign({}, this.options);
        this._options.element = this.elementRef.nativeElement;
        this._options.data = this.data;
    }
    ngAfterViewInit() {
        if (!this.window.Morris) {
            throw new Error('Please include nodes_modules/morris.js/morris.js');
        }
        else {
            this.ngZone.runOutsideAngular(() => {
                this.chartInstance = new this.window.Morris[this.type](this._options);
            });
        }
    }
    ngOnChanges(changes) {
        if (changes["data"] && !changes["data"].firstChange) {
            this.chartInstance.setData(this.data);
        }
    }
    ngOnDestroy() {
        if (this.chartInstance.el.empty instanceof Function) {
            this.chartInstance.el.empty();
        }
    }
}
MorrisChartDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mk-morris-js]'
            },] },
];
MorrisChartDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: NgZone, },
];
MorrisChartDirective.propDecorators = {
    "type": [{ type: Input },],
    "options": [{ type: Input },],
    "data": [{ type: Input },],
};

class MorrisJsModule {
}
MorrisJsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [MorrisChartDirective],
                exports: [MorrisChartDirective]
            },] },
];

export { MorrisChartDirective, MorrisJsModule };
