// angular-morris-js/esm5/angular-morris-js.js
import { Directive, Input, ElementRef, NgZone, NgModule } from '@angular/core';

var MorrisChartDirective = /** @class */ (function () {
    function MorrisChartDirective(elementRef, ngZone) {
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.window = window;
        this.type = 'Line';
    }
    MorrisChartDirective.prototype.ngOnInit = function () {
        this._options = Object.assign({}, this.options);
        this._options.element = this.elementRef.nativeElement;
        this._options.data = this.data;
    };
    MorrisChartDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (!this.window.Morris) {
            throw new Error('Please include nodes_modules/morris.js/morris.js');
        }
        else {
            this.ngZone.runOutsideAngular(function () {
                _this.chartInstance = new _this.window.Morris[_this.type](_this._options);
            });
        }
    };
    MorrisChartDirective.prototype.ngOnChanges = function (changes) {
        if (changes["data"] && !changes["data"].firstChange) {
            this.chartInstance.setData(this.data);
        }
    };
    MorrisChartDirective.prototype.ngOnDestroy = function () {
        if (this.chartInstance.el.empty instanceof Function) {
            this.chartInstance.el.empty();
        }
    };
    return MorrisChartDirective;
}());
MorrisChartDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mk-morris-js]'
            },] },
];
MorrisChartDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: NgZone, },
]; };
MorrisChartDirective.propDecorators = {
    "type": [{ type: Input },],
    "options": [{ type: Input },],
    "data": [{ type: Input },],
};
var MorrisJsModule = /** @class */ (function () {
    function MorrisJsModule() {
    }
    return MorrisJsModule;
}());
MorrisJsModule.decorators = [
    { type: NgModule, args: [{
                declarations: [MorrisChartDirective],
                exports: [MorrisChartDirective]
            },] },
];

export { MorrisChartDirective, MorrisJsModule };
