// angular-highcharts-chart/fesm5/angular-highcharts-chart.js
import { Component, ElementRef, EventEmitter, Input, Output, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AngularHighchartsChartComponent = /** @class */ (function () {
    function AngularHighchartsChartComponent(el) {
        this.el = el;
        this.updateValue = false;
        this.updateChange = new EventEmitter(true);
        this.updateOrCreateChart = function () {
            if (this.chart && this.chart.update) {
                this.chart.update(this.optionsValue, true, this.oneToOne || false);
            }
            else {
                this.chart = this.Highcharts[this.constructorType || 'chart'](this.el.nativeElement, this.optionsValue, this.callbackFunction || null);
                this.optionsValue.series = this.chart.userOptions.series;
            }
        };
    }
    Object.defineProperty(AngularHighchartsChartComponent.prototype, "options", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this.optionsValue = val;
            this.updateOrCreateChart();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AngularHighchartsChartComponent.prototype, "update", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val) {
                this.updateOrCreateChart();
                this.updateChange.emit(false); // clear the flag after update
            }
        },
        enumerable: true,
        configurable: true
    });
    AngularHighchartsChartComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-highcharts-angular',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    AngularHighchartsChartComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AngularHighchartsChartComponent.propDecorators = {
        Highcharts: [{ type: Input }],
        constructorType: [{ type: Input }],
        callbackFunction: [{ type: Input }],
        options: [{ type: Input }],
        updateChange: [{ type: Output }],
        update: [{ type: Input }],
        oneToOne: [{ type: Input }]
    };
    return AngularHighchartsChartComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AngularHighchartsChartModule = /** @class */ (function () {
    function AngularHighchartsChartModule() {
    }
    AngularHighchartsChartModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [AngularHighchartsChartComponent],
                    exports: [AngularHighchartsChartComponent]
                },] },
    ];
    return AngularHighchartsChartModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// export * from './lib/angular-highcharts-chart.component';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { AngularHighchartsChartModule, AngularHighchartsChartComponent as ɵa };
