// angular2-tinymce/esm5/lib/angular2-tinymce-lib.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { TinymceComponent } from './angular2-tinymce-lib.component';
import { TinymceDefaultOptions } from './angular2-tinymce-lib.default';
var TinymceModule = /** @class */ (function () {
    function TinymceModule() {
    }
    /**
     * @param {?=} userConfig
     * @return {?}
     */
    TinymceModule.withConfig = /**
     * @param {?=} userConfig
     * @return {?}
     */
    function (userConfig) {
        if (userConfig === void 0) { userConfig = {}; }
        return {
            ngModule: TinymceModule,
            providers: [
                { provide: 'TINYMCE_CONFIG', useValue: userConfig }
            ]
        };
    };
    TinymceModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [TinymceComponent],
                    exports: [TinymceComponent],
                    providers: [
                        { provide: 'TINYMCE_CONFIG', useClass: TinymceDefaultOptions }
                    ]
                },] }
    ];
    return TinymceModule;
}());
export { TinymceModule };
