// angular-morris-js/bundles/angular-morris-js.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
	typeof define === 'function' && define.amd ? define('angular-morris-js', ['exports', '@angular/core'], factory) :
	(factory((global['angular-morris-js'] = {}),global.ng.core));
}(this, (function (exports,core) { 'use strict';

var MorrisChartDirective =               (function () {
    function MorrisChartDirective(elementRef, ngZone) {
        this.elementRef = elementRef;
        this.ngZone = ngZone;
        this.window = window;
        this.type = 'Line';
    }
    MorrisChartDirective.prototype.ngOnInit = function () {
        this._options = Object.assign({}, this.options);
        this._options.element = this.elementRef.nativeElement;
        this._options.data = this.data;
    };
    MorrisChartDirective.prototype.ngAfterViewInit = function () {
        var _this = this;
        if (!this.window.Morris) {
            throw new Error('Please include nodes_modules/morris.js/morris.js');
        }
        else {
            this.ngZone.runOutsideAngular(function () {
                _this.chartInstance = new _this.window.Morris[_this.type](_this._options);
            });
        }
    };
    MorrisChartDirective.prototype.ngOnChanges = function (changes) {
        if (changes["data"] && !changes["data"].firstChange) {
            this.chartInstance.setData(this.data);
        }
    };
    MorrisChartDirective.prototype.ngOnDestroy = function () {
        if (this.chartInstance.el.empty instanceof Function) {
            this.chartInstance.el.empty();
        }
    };
    return MorrisChartDirective;
}());
MorrisChartDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[mk-morris-js]'
            },] },
];
MorrisChartDirective.ctorParameters = function () { return [
    { type: core.ElementRef, },
    { type: core.NgZone, },
]; };
MorrisChartDirective.propDecorators = {
    "type": [{ type: core.Input },],
    "options": [{ type: core.Input },],
    "data": [{ type: core.Input },],
};
var MorrisJsModule =               (function () {
    function MorrisJsModule() {
    }
    return MorrisJsModule;
}());
MorrisJsModule.decorators = [
    { type: core.NgModule, args: [{
                declarations: [MorrisChartDirective],
                exports: [MorrisChartDirective]
            },] },
];

exports.MorrisChartDirective = MorrisChartDirective;
exports.MorrisJsModule = MorrisJsModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
