// @nicky-lenaers/ngx-scroll-to/esm2015/src/app/modules/scroll-to/scroll-to-config.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ScrollToConfigOptionsBase() { }
function ScrollToConfigOptionsBase_tsickle_Closure_declarations() {
    /**
     * The Container to scroll
     * @type {?|undefined}
     */
    ScrollToConfigOptionsBase.prototype.container;
    /**
     * Duration of the Scroll Animation
     * @type {?|undefined}
     */
    ScrollToConfigOptionsBase.prototype.duration;
    /**
     * The named Easing Function to use
     * @type {?|undefined}
     */
    ScrollToConfigOptionsBase.prototype.easing;
    /**
     * A mapped list of offsets
     * @type {?|undefined}
     */
    ScrollToConfigOptionsBase.prototype.offsetMap;
}
/**
 * Configuration Options Target
 * @record
 */
export function ScrollToConfigOptionsTarget() { }
function ScrollToConfigOptionsTarget_tsickle_Closure_declarations() {
    /**
     * The target to scroll to
     * @type {?}
     */
    ScrollToConfigOptionsTarget.prototype.target;
    /**
     * The offset from the top of the Element in pixels
     * @type {?|undefined}
     */
    ScrollToConfigOptionsTarget.prototype.offset;
}
/**
 * Configuration Options Offset Only
 * @record
 */
export function ScrollToConfigOptionsOffsetOnly() { }
function ScrollToConfigOptionsOffsetOnly_tsickle_Closure_declarations() {
    /**
     * The offset from the top of the Element in pixels
     * @type {?}
     */
    ScrollToConfigOptionsOffsetOnly.prototype.offset;
}
/**
 * The Default Configuration Object
 * @record
 */
export function ScrollToDefaultConfigOptions() { }
function ScrollToDefaultConfigOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    ScrollToDefaultConfigOptions.prototype.action;
}
