// ng2-material-dropdown/bundles/ng2-material-dropdown.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/animations'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng2-material-dropdown', ['exports', '@angular/core', '@angular/animations', '@angular/common'], factory) :
    (factory((global['ng2-material-dropdown'] = {}),global.ng.core,global.ng.animations,global.ng.common));
}(this, (function (exports,core,animations,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Ng2DropdownButton = (function () {
        function Ng2DropdownButton(element) {
            this.element = element;
            this.onMenuToggled = new core.EventEmitter();
            this.showCaret = true;
        }
        /**
         * \@name toggleMenu
         * @desc emits event to toggle menu
         * @return {?}
         */
        Ng2DropdownButton.prototype.toggleMenu = /**
         * \@name toggleMenu
         * @desc emits event to toggle menu
         * @return {?}
         */
            function () {
                this.onMenuToggled.emit(true);
            };
        /**
         * \@name getPosition
         * @desc returns position of the button
         * @return {?}
         */
        Ng2DropdownButton.prototype.getPosition = /**
         * \@name getPosition
         * @desc returns position of the button
         * @return {?}
         */
            function () {
                return this.element.nativeElement.getBoundingClientRect();
            };
        Ng2DropdownButton.decorators = [
            { type: core.Component, args: [{
                        selector: 'ng2-dropdown-button',
                        styles: [".ng2-dropdown-button{font-family:Roboto,\"Helvetica Neue\",Helvetica,Arial;background:#fff;padding:.45rem .25rem;font-size:14px;letter-spacing:.08rem;color:#444;outline:0;cursor:pointer;font-weight:400;border:none;border-bottom:1px solid #efefef;text-align:left;min-width:100px;width:100%;display:flex;flex-direction:row;max-width:150px}.ng2-dropdown-button:hover{color:#222}.ng2-dropdown-button:active,.ng2-dropdown-button:focus{color:#222;border-bottom:2px solid #2196f3}.ng2-dropdown-button__label{flex:1 1 95%}.ng2-dropdown-button__caret{width:12px;height:12px;display:flex;flex:1 1 6%}:host-context(.ng2-dropdown-button--icon) .ng2-dropdown-button{border:none;min-width:40px;width:40px;border-radius:100%;transition:all .2s;text-align:center;height:40px;padding:.5em}:host-context(.ng2-dropdown-button--icon) .ng2-dropdown-button:active{background:rgba(0,0,0,.2)}"],
                        template: "<button class='ng2-dropdown-button' type=\"button\" (click)=\"toggleMenu()\" tabindex=\"0s\">\n    <span class=\"ng2-dropdown-button__label\">\n        <ng-content></ng-content>\n    </span>\n\n    <span class=\"ng2-dropdown-button__caret\" *ngIf=\"showCaret\">\n        <svg enable-background=\"new 0 0 32 32\" height=\"16px\" id=\"\u0421\u043B\u043E\u0439_1\" version=\"1.1\" viewBox=\"0 0 32 32\" width=\"16px\" xml:space=\"preserve\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><path d=\"M24.285,11.284L16,19.571l-8.285-8.288c-0.395-0.395-1.034-0.395-1.429,0  c-0.394,0.395-0.394,1.035,0,1.43l8.999,9.002l0,0l0,0c0.394,0.395,1.034,0.395,1.428,0l8.999-9.002  c0.394-0.395,0.394-1.036,0-1.431C25.319,10.889,24.679,10.889,24.285,11.284z\" fill=\"#121313\" id=\"Expand_More\"/><g/><g/><g/><g/><g/><g/></svg>\n    </span>\n</button>\n"
                    },] },
        ];
        /** @nocollapse */
        Ng2DropdownButton.ctorParameters = function () {
            return [
                { type: core.ElementRef, },
            ];
        };
        Ng2DropdownButton.propDecorators = {
            "onMenuToggled": [{ type: core.Output },],
            "showCaret": [{ type: core.Input },],
        };
        return Ng2DropdownButton;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ KEYS = {
        BACKSPACE: 9,
        PREV: 38,
        NEXT: 40,
        ENTER: 13,
        ESCAPE: 27
    };
    /**
     * \@name onSwitchNext
     * @param index
     * @param items
     * @param state
     */
    var /** @type {?} */ onSwitchNext = function (index, items, state) {
        if (index < items.length - 1) {
            state.select(items[index + 1], true);
        }
    };
    /**
     * \@name onSwitchPrev
     * @param index
     * @param items
     * @param state
     */
    var /** @type {?} */ onSwitchPrev = function (index, items, state) {
        if (index > 0) {
            state.select(items[index - 1], true);
        }
    };
    /**
     * \@name onBackspace
     * @param index
     * @param items
     * @param state
     */
    var /** @type {?} */ onBackspace = function (index, items, state) {
        if (index < items.length - 1) {
            state.select(items[index + 1], true);
        }
        else {
            state.select(items[0], true);
        }
    };
    /**
     * @this {?}
     * @return {?}
     */
    function onEscape() {
        this.hide();
    }
    /**
     * \@name onItemClicked
     * @param index
     * @param items
     * @param state
     */
    var /** @type {?} */ onItemClicked = function (index, items, state) {
        return state.selectedItem ? state.selectedItem.click() : undefined;
    };
    var /** @type {?} */ ACTIONS = (_a = {},
        _a[KEYS.BACKSPACE] = onBackspace,
        _a[KEYS.PREV] = onSwitchPrev,
        _a[KEYS.NEXT] = onSwitchNext,
        _a[KEYS.ENTER] = onItemClicked,
        _a[KEYS.ESCAPE] = onEscape,
        _a);
    /**
     * @param {?} event
     * @return {?}
     */
    function arrowKeysHandler(event) {
        if ([38, 40].indexOf(event.keyCode) > -1) {
            event.preventDefault();
        }
    }
    var _a;

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Ng2DropdownState = (function () {
        function Ng2DropdownState() {
            this.onItemSelected = new core.EventEmitter();
            this.onItemClicked = new core.EventEmitter();
            this.onItemDestroyed = new core.EventEmitter();
        }
        Object.defineProperty(Ng2DropdownState.prototype, "selectedItem", {
            get: /**
             * \@name selectedItem
             * @desc getter for _selectedItem
             * @return {?}
             */ function () {
                return this._selectedItem;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * \@name selects a menu item and emits event
         * @param {?} item
         * @param {?=} dispatchEvent
         * @return {?}
         */
        Ng2DropdownState.prototype.select = /**
         * \@name selects a menu item and emits event
         * @param {?} item
         * @param {?=} dispatchEvent
         * @return {?}
         */
            function (item, dispatchEvent) {
                if (dispatchEvent === void 0) {
                    dispatchEvent = true;
                }
                this._selectedItem = item;
                if (!dispatchEvent || !item) {
                    return;
                }
                item.focus();
                this.onItemSelected.emit(item);
            };
        /**
         * \@name unselect
         * @desc sets _selectedItem as undefined
         * @return {?}
         */
        Ng2DropdownState.prototype.unselect = /**
         * \@name unselect
         * @desc sets _selectedItem as undefined
         * @return {?}
         */
            function () {
                this._selectedItem = undefined;
            };
        return Ng2DropdownState;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var DropdownStateService = (function () {
        function DropdownStateService() {
            this.menuState = {
                isVisible: /** @type {?} */ (false),
                toString: /**
                 * @return {?}
                 */ function () {
                    return this.isVisible === true ? 'visible' : 'hidden';
                }
            };
            this.dropdownState = new Ng2DropdownState();
        }
        DropdownStateService.decorators = [
            { type: core.Injectable },
        ];
        return DropdownStateService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Ng2MenuItem = (function () {
        function Ng2MenuItem(state, element, renderer) {
            this.state = state;
            this.element = element;
            this.renderer = renderer;
            /**
             * \@preventClose
             * @desc if true, clicking on the item won't close the dropdown
             */
            this.preventClose = false;
        }
        /**
         * @return {?}
         */
        Ng2MenuItem.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this.state.dropdownState.onItemDestroyed.emit(this);
            };
        Object.defineProperty(Ng2MenuItem.prototype, "isSelected", {
            get: /**
             * \@name isSelected
             * @desc returns current selected item
             * @return {?}
             */ function () {
                return this === this.state.dropdownState.selectedItem;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * \@name click
         * @desc emits select event
         * @param {?=} $event
         * @return {?}
         */
        Ng2MenuItem.prototype.select = /**
         * \@name click
         * @desc emits select event
         * @param {?=} $event
         * @return {?}
         */
            function ($event) {
                this.state.dropdownState.select(this, true);
                if ($event) {
                    $event.stopPropagation();
                    $event.preventDefault();
                }
            };
        /**
         * \@name click
         * @desc emits click event
         * @return {?}
         */
        Ng2MenuItem.prototype.click = /**
         * \@name click
         * @desc emits click event
         * @return {?}
         */
            function () {
                this.state.dropdownState.onItemClicked.emit(this);
            };
        /**
         * \@name focus
         * @return {?}
         */
        Ng2MenuItem.prototype.focus = /**
         * \@name focus
         * @return {?}
         */
            function () {
                this.renderer.invokeElementMethod(this.element.nativeElement.children[0], 'focus');
            };
        Ng2MenuItem.decorators = [
            { type: core.Component, args: [{
                        selector: 'ng2-menu-item',
                        styles: [".ng2-menu-item{font-family:Roboto,\"Helvetica Neue\",Helvetica,Arial;background:#fff;color:rgba(0,0,0,.87);cursor:pointer;font-size:.9em;text-transform:none;font-weight:400;letter-spacing:.03em;height:48px;line-height:48px;padding:.3em 1.25rem;text-overflow:ellipsis;white-space:nowrap;overflow:hidden;transition:background .25s}.ng2-menu-item--selected{background:rgba(158,158,158,.2);outline:0}.ng2-menu-item:focus{outline:0}.ng2-menu-item:active{background:rgba(158,158,158,.4)}:host(ng2-menu-item) /deep/ [ng2-menu-item-icon]{vertical-align:middle;font-size:28px;width:1.5em;height:30px;color:rgba(0,0,0,.44)}"],
                        template: "<div class='ng2-menu-item'\n     role=\"button\"\n     tabindex=\"0\"\n     [class.ng2-menu-item--selected]=\"isSelected\"\n     (keydown.enter)=\"click()\"\n     (click)=\"click()\"\n     (mouseover)=\"select()\">\n        <ng-content></ng-content>\n</div>\n"
                    },] },
        ];
        /** @nocollapse */
        Ng2MenuItem.ctorParameters = function () {
            return [
                { type: DropdownStateService, },
                { type: core.ElementRef, },
                { type: core.Renderer, },
            ];
        };
        Ng2MenuItem.propDecorators = {
            "preventClose": [{ type: core.Input },],
            "value": [{ type: core.Input },],
        };
        return Ng2MenuItem;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Ng2DropdownMenu = (function () {
        function Ng2DropdownMenu(state, element, renderer) {
            this.state = state;
            this.element = element;
            this.renderer = renderer;
            /**
             * \@name width
             */
            this.width = 4;
            /**
             * \@description if set to true, the first element of the dropdown will be automatically focused
             * \@name focusFirstElement
             */
            this.focusFirstElement = true;
            /**
             * \@name appendToBody
             */
            this.appendToBody = true;
            /**
             * \@name zIndex
             */
            this.zIndex = 1000;
            this.listeners = {
                arrowHandler: undefined,
                handleKeypress: undefined
            };
        }
        /**
         * \@name show
         * \@shows menu and selects first item
         * @param {?=} position
         * @param {?=} dynamic
         * @return {?}
         */
        Ng2DropdownMenu.prototype.show = /**
         * \@name show
         * \@shows menu and selects first item
         * @param {?=} position
         * @param {?=} dynamic
         * @return {?}
         */
            function (position, dynamic) {
                if (dynamic === void 0) {
                    dynamic = true;
                }
                var /** @type {?} */ dc = typeof document !== 'undefined' ? document : undefined;
                var /** @type {?} */ wd = typeof window !== 'undefined' ? window : undefined;
                if (!this.state.menuState.isVisible) {
                    // setting handlers
                    this.listeners.handleKeypress = this.renderer.listen(dc.body, 'keydown', this.handleKeypress.bind(this));
                    this.listeners.arrowHandler = this.renderer.listen(wd, 'keydown', arrowKeysHandler);
                }
                // update state
                this.state.menuState.isVisible = true;
                if (position) {
                    this.updatePosition(position, dynamic);
                }
            };
        /**
         * \@name hide
         * @desc hides menu
         * @return {?}
         */
        Ng2DropdownMenu.prototype.hide = /**
         * \@name hide
         * @desc hides menu
         * @return {?}
         */
            function () {
                this.state.menuState.isVisible = false;
                // reset selected item state
                this.state.dropdownState.unselect();
                // call function to unlisten
                this.listeners.arrowHandler ? this.listeners.arrowHandler() : undefined;
                this.listeners.handleKeypress ? this.listeners.handleKeypress() : undefined;
            };
        /**
         * \@name updatePosition
         * @desc updates the menu position every time it is toggled
         * @param {?} position {ClientRect}
         * @param {?} dynamic {boolean}
         * @return {?}
         */
        Ng2DropdownMenu.prototype.updatePosition = /**
         * \@name updatePosition
         * @desc updates the menu position every time it is toggled
         * @param {?} position {ClientRect}
         * @param {?} dynamic {boolean}
         * @return {?}
         */
            function (position, dynamic) {
                this.position = position;
                this.updateOnChange(dynamic);
            };
        /**
         * \@name handleKeypress
         * @desc executes functions on keyPress based on the key pressed
         * @param {?} $event
         * @return {?}
         */
        Ng2DropdownMenu.prototype.handleKeypress = /**
         * \@name handleKeypress
         * @desc executes functions on keyPress based on the key pressed
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                var /** @type {?} */ key = $event.keyCode;
                var /** @type {?} */ items = this.items.toArray();
                var /** @type {?} */ index = items.indexOf(this.state.dropdownState.selectedItem);
                if (!ACTIONS.hasOwnProperty(key)) {
                    return;
                }
                ACTIONS[key].call(this, index, items, this.state.dropdownState);
            };
        /**
         * \@name getMenuElement
         * @return {?}
         */
        Ng2DropdownMenu.prototype.getMenuElement = /**
         * \@name getMenuElement
         * @return {?}
         */
            function () {
                return this.element.nativeElement.children[0];
            };
        /**
         * \@name calcPositionOffset
         * @param {?} position
         * @return {?}
         */
        Ng2DropdownMenu.prototype.calcPositionOffset = /**
         * \@name calcPositionOffset
         * @param {?} position
         * @return {?}
         */
            function (position) {
                var /** @type {?} */ wd = typeof window !== 'undefined' ? window : undefined;
                var /** @type {?} */ dc = typeof document !== 'undefined' ? document : undefined;
                if (!wd || !dc || !position) {
                    return;
                }
                var /** @type {?} */ element = this.getMenuElement();
                var /** @type {?} */ supportPageOffset = wd.pageXOffset !== undefined;
                var /** @type {?} */ isCSS1Compat = ((dc.compatMode || '') === 'CSS1Compat');
                var /** @type {?} */ x = supportPageOffset ? wd.pageXOffset : isCSS1Compat ?
                    dc.documentElement.scrollLeft : dc.body.scrollLeft;
                var /** @type {?} */ y = supportPageOffset ? wd.pageYOffset : isCSS1Compat ?
                    dc.documentElement.scrollTop : dc.body.scrollTop;
                var _a = this.applyOffset(position.top + (this.appendToBody ? y - 15 : 0) + "px", position.left + x - 5 + "px"), top = _a.top, left = _a.left;
                var /** @type {?} */ clientWidth = element.clientWidth;
                var /** @type {?} */ clientHeight = element.clientHeight;
                var /** @type {?} */ marginFromBottom = parseInt(top) + clientHeight + (this.appendToBody ? 0 : y - 15);
                var /** @type {?} */ marginFromRight = parseInt(left) + clientWidth;
                var /** @type {?} */ windowScrollHeight = wd.innerHeight + wd.scrollY;
                var /** @type {?} */ windowScrollWidth = wd.innerWidth + wd.scrollX;
                if (marginFromBottom >= windowScrollHeight) {
                    top = parseInt(top.replace('px', '')) - clientHeight + "px";
                }
                if (marginFromRight >= windowScrollWidth) {
                    var /** @type {?} */ marginRight = marginFromRight - windowScrollWidth + 30;
                    left = parseInt(left.replace('px', '')) - marginRight + "px";
                }
                return { top: top, left: left };
            };
        /**
         * @param {?} top
         * @param {?} left
         * @return {?}
         */
        Ng2DropdownMenu.prototype.applyOffset = /**
         * @param {?} top
         * @param {?} left
         * @return {?}
         */
            function (top, left) {
                if (!this.offset) {
                    return { top: top, left: left };
                }
                var /** @type {?} */ offset = this.offset.split(' ');
                if (!offset[1]) {
                    offset[1] = '0';
                }
                top = parseInt(top.replace('px', '')) + parseInt(offset[0]) + "px";
                left = parseInt(left.replace('px', '')) + parseInt(offset[1]) + "px";
                return { top: top, left: left };
            };
        /**
         * @return {?}
         */
        Ng2DropdownMenu.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ dc = typeof document !== 'undefined' ? document : undefined;
                if (this.appendToBody) {
                    // append menu element to the body
                    dc.body.appendChild(this.element.nativeElement);
                }
            };
        /**
         * @param {?=} dynamic
         * @return {?}
         */
        Ng2DropdownMenu.prototype.updateOnChange = /**
         * @param {?=} dynamic
         * @return {?}
         */
            function (dynamic) {
                if (dynamic === void 0) {
                    dynamic = true;
                }
                var /** @type {?} */ element = this.getMenuElement();
                var /** @type {?} */ position = this.calcPositionOffset(this.position);
                if (position) {
                    this.renderer.setElementStyle(element, 'top', position.top.toString());
                    this.renderer.setElementStyle(element, 'left', position.left.toString());
                }
                // select first item unless user disabled this option
                if (this.focusFirstElement &&
                    this.items.first &&
                    !this.state.dropdownState.selectedItem) {
                    this.state.dropdownState.select(this.items.first, false);
                }
            };
        /**
         * @return {?}
         */
        Ng2DropdownMenu.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ elem = this.element.nativeElement;
                elem.parentNode.removeChild(elem);
                if (this.listeners.handleKeypress) {
                    this.listeners.handleKeypress();
                }
            };
        Ng2DropdownMenu.decorators = [
            { type: core.Component, args: [{
                        selector: 'ng2-dropdown-menu',
                        styles: [":host{display:block}.ng2-dropdown-menu{overflow-y:auto;box-shadow:0 1px 2px 0 rgba(0,0,0,.3);position:absolute;padding:.5em 0;background:#fff;border-radius:1px;max-height:400px;width:260px;min-height:0;display:block}.ng2-dropdown-menu.ng2-dropdown-menu--inside-element{position:fixed}.ng2-dropdown-menu.ng2-dropdown-menu--width--2{width:200px}.ng2-dropdown-menu.ng2-dropdown-menu--width--4{width:260px}.ng2-dropdown-menu.ng2-dropdown-menu--width--6{width:320px}.ng2-dropdown-backdrop{position:fixed;top:0;left:0;width:100%;height:100%;z-index:1;overflow:hidden}:host /deep/ .ng2-menu-divider{height:1px;min-height:1px;max-height:1px;width:100%;display:block;background:#f9f9f9}"],
                        template: "<!-- MENU -->\n<div class='ng2-dropdown-menu ng2-dropdown-menu---width--{{ width }}'\n     [class.ng2-dropdown-menu--inside-element]=\"!appendToBody\"\n     [class.ng2-dropdown-menu--open]=\"state.menuState.isVisible\"\n     [style.z-index]=\"zIndex\"\n     [@fade]=\"state.menuState.toString()\">\n        <div class=\"ng2-dropdown-menu__options-container\"\n             [@opacity]=\"state.menuState.toString()\">\n            <ng-content></ng-content>\n        </div>\n</div>\n\n<!-- BACKDROP -->\n<div class=\"ng2-dropdown-backdrop\" *ngIf=\"state.menuState.isVisible\" (click)=\"hide()\"></div>\n",
                        animations: [
                            animations.trigger('fade', [
                                animations.state('visible', animations.style({ display: 'block', opacity: 1, height: '*', width: '*' })),
                                animations.state('hidden', animations.style({ display: 'none', opacity: 0, overflow: 'hidden', height: 0, width: 0 })),
                                animations.transition('hidden => visible', [
                                    animations.animate('250ms ease-in', animations.style({ opacity: 1, height: '*', width: '*' }))
                                ]),
                                animations.transition('visible => hidden', [
                                    animations.animate('350ms ease-out', animations.style({ opacity: 0, width: 0, height: 0 }))
                                ])
                            ]),
                            animations.trigger('opacity', [
                                animations.transition('hidden => visible', [
                                    animations.animate('450ms ease-in', animations.keyframes([
                                        animations.style({ opacity: 0, offset: 0 }),
                                        animations.style({ opacity: 1, offset: 1 }),
                                    ]))
                                ]),
                                animations.transition('visible => hidden', [
                                    animations.animate('250ms ease-out', animations.keyframes([
                                        animations.style({ opacity: 1, offset: 0 }),
                                        animations.style({ opacity: 0.5, offset: 0.3 }),
                                        animations.style({ opacity: 0, offset: 1 }),
                                    ]))
                                ])
                            ])
                        ]
                    },] },
        ];
        /** @nocollapse */
        Ng2DropdownMenu.ctorParameters = function () {
            return [
                { type: DropdownStateService, },
                { type: core.ElementRef, },
                { type: core.Renderer, },
            ];
        };
        Ng2DropdownMenu.propDecorators = {
            "width": [{ type: core.Input },],
            "focusFirstElement": [{ type: core.Input },],
            "offset": [{ type: core.Input },],
            "appendToBody": [{ type: core.Input },],
            "zIndex": [{ type: core.Input },],
            "items": [{ type: core.ContentChildren, args: [Ng2MenuItem,] },],
        };
        return Ng2DropdownMenu;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Ng2Dropdown = (function () {
        function Ng2Dropdown(state) {
            this.state = state;
            this.dynamicUpdate = true;
            // outputs
            this.onItemClicked = new core.EventEmitter();
            this.onItemSelected = new core.EventEmitter();
            this.onShow = new core.EventEmitter();
            this.onHide = new core.EventEmitter();
        }
        /**
         * @return {?}
         */
        Ng2Dropdown.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                this.state.dropdownState.onItemClicked.subscribe(function (item) {
                    _this.onItemClicked.emit(item);
                    if (item.preventClose) {
                        return;
                    }
                    _this.hide.call(_this);
                });
                if (this.button) {
                    this.button.onMenuToggled.subscribe(function () {
                        _this.toggleMenu();
                    });
                }
                this.state.dropdownState.onItemSelected.subscribe(function (item) {
                    _this.onItemSelected.emit(item);
                });
                this.state.dropdownState.onItemDestroyed.subscribe(function (item) {
                    var /** @type {?} */ newSelectedItem;
                    var /** @type {?} */ items = _this.menu.items.toArray();
                    if (item !== _this.state.dropdownState.selectedItem) {
                        return;
                    }
                    if (_this.menu.focusFirstElement) {
                        newSelectedItem = item === items[0] && items.length > 1 ? items[1] : items[0];
                    }
                    _this.state.dropdownState.select(newSelectedItem);
                });
            };
        /**
         * \@name toggleMenu
         * @desc toggles menu visibility
         * @param {?=} position
         * @return {?}
         */
        Ng2Dropdown.prototype.toggleMenu = /**
         * \@name toggleMenu
         * @desc toggles menu visibility
         * @param {?=} position
         * @return {?}
         */
            function (position) {
                if (position === void 0) {
                    position = this.button.getPosition();
                }
                this.state.menuState.isVisible ? this.hide() : this.show(position);
            };
        /**
         * - hides dropdown
         * \@name hide
         * @return {?}
         */
        Ng2Dropdown.prototype.hide = /**
         * - hides dropdown
         * \@name hide
         * @return {?}
         */
            function () {
                this.menu.hide();
                this.onHide.emit(this);
            };
        /**
         * - shows dropdown
         * \@name show
         * @param {?=} position
         * @return {?}
         */
        Ng2Dropdown.prototype.show = /**
         * - shows dropdown
         * \@name show
         * @param {?=} position
         * @return {?}
         */
            function (position) {
                if (position === void 0) {
                    position = this.button.getPosition();
                }
                this.menu.show(position, this.dynamicUpdate);
                this.onShow.emit(this);
            };
        /**
         * \@name scrollListener
         * @return {?}
         */
        Ng2Dropdown.prototype.scrollListener = /**
         * \@name scrollListener
         * @return {?}
         */
            function () {
                if (this.button && this.dynamicUpdate) {
                    this.menu.updatePosition(this.button.getPosition(), true);
                }
            };
        Ng2Dropdown.decorators = [
            { type: core.Component, args: [{
                        selector: 'ng2-dropdown',
                        template: "<div class=\"ng2-dropdown-container\">\n    <ng-content select=\"ng2-dropdown-button\"></ng-content>\n    <ng-content select=\"ng2-dropdown-menu\"></ng-content>\n</div>\n",
                        providers: [DropdownStateService]
                    },] },
        ];
        /** @nocollapse */
        Ng2Dropdown.ctorParameters = function () {
            return [
                { type: DropdownStateService, },
            ];
        };
        Ng2Dropdown.propDecorators = {
            "button": [{ type: core.ContentChild, args: [Ng2DropdownButton,] },],
            "menu": [{ type: core.ContentChild, args: [Ng2DropdownMenu,] },],
            "dynamicUpdate": [{ type: core.Input },],
            "onItemClicked": [{ type: core.Output },],
            "onItemSelected": [{ type: core.Output },],
            "onShow": [{ type: core.Output },],
            "onHide": [{ type: core.Output },],
            "scrollListener": [{ type: core.HostListener, args: ['window:scroll',] },],
        };
        return Ng2Dropdown;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Ng2DropdownModule = (function () {
        function Ng2DropdownModule() {
        }
        Ng2DropdownModule.decorators = [
            { type: core.NgModule, args: [{
                        exports: [
                            Ng2MenuItem,
                            Ng2DropdownButton,
                            Ng2DropdownMenu,
                            Ng2Dropdown
                        ],
                        declarations: [
                            Ng2Dropdown,
                            Ng2MenuItem,
                            Ng2DropdownButton,
                            Ng2DropdownMenu,
                        ],
                        imports: [
                            common.CommonModule
                        ]
                    },] },
        ];
        return Ng2DropdownModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.Ng2Dropdown = Ng2Dropdown;
    exports.Ng2DropdownMenu = Ng2DropdownMenu;
    exports.Ng2MenuItem = Ng2MenuItem;
    exports.Ng2DropdownButton = Ng2DropdownButton;
    exports.Ng2DropdownModule = Ng2DropdownModule;
    exports.DropdownStateService = DropdownStateService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
