// ng2-material-dropdown/esm5/src/modules/services/dropdown-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Ng2DropdownState } from './ng2-dropdown-state';
var DropdownStateService = /** @class */ (function () {
    function DropdownStateService() {
        this.menuState = {
            isVisible: /** @type {?} */ (false),
            toString: /**
             * @return {?}
             */
            function () {
                return this.isVisible === true ? 'visible' : 'hidden';
            }
        };
        this.dropdownState = new Ng2DropdownState();
    }
    DropdownStateService.decorators = [
        { type: Injectable },
    ];
    return DropdownStateService;
}());
export { DropdownStateService };
function DropdownStateService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DropdownStateService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DropdownStateService.ctorParameters;
    /** @type {?} */
    DropdownStateService.prototype.menuState;
    /** @type {?} */
    DropdownStateService.prototype.dropdownState;
}
