// ng-number-formatter/bundles/ng-number-formatter.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-number-formatter', ['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global['ng-number-formatter'] = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgNumberFormatterComponent = (function () {
        function NgNumberFormatterComponent() {
        }
        /**
         * @return {?}
         */
        NgNumberFormatterComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                /** @type {?} */
                var incomingNumber = this.ngNumber;
                this.ngNumber = this.nFormatter(Math.abs(parseInt(incomingNumber)), 1);
            };
        /**
         * @param {?} num
         * @param {?} digits
         * @return {?}
         */
        NgNumberFormatterComponent.prototype.nFormatter = /**
         * @param {?} num
         * @param {?} digits
         * @return {?}
         */
            function (num, digits) {
                /** @type {?} */
                var si = [
                    { value: 1E18, symbol: "E" },
                    { value: 1E15, symbol: "P" },
                    { value: 1E12, symbol: "T" },
                    { value: 1E9, symbol: "B" },
                    { value: 1E6, symbol: "M" },
                    { value: 1E3, symbol: "K" }
                ];
                /** @type {?} */
                var rx = /\.0+$|(\.[0-9]*[1-9])0+$/;
                /** @type {?} */
                var i;
                for (i = 0; i < si.length; i++) {
                    if (num >= si[i].value) {
                        return (num / si[i].value).toFixed(digits).replace(rx, "$1") + si[i].symbol;
                    }
                }
                return num.toFixed(digits).replace(rx, "$1");
            };
        NgNumberFormatterComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ng-number-formatter',
                        inputs: ['ngNumber'],
                        template: '<span>{{ngNumber}}</span>',
                    },] },
        ];
        /** @nocollapse */
        NgNumberFormatterComponent.ctorParameters = function () { return []; };
        return NgNumberFormatterComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgNumberFormatterModule = (function () {
        function NgNumberFormatterModule() {
        }
        NgNumberFormatterModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            NgNumberFormatterComponent
                        ],
                        exports: [
                            NgNumberFormatterComponent
                        ]
                    },] },
        ];
        return NgNumberFormatterModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.NgNumberFormatterModule = NgNumberFormatterModule;
    exports.ɵa = NgNumberFormatterComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
