// ng2-material-dropdown/esm2015/src/modules/services/dropdown-state.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Ng2DropdownState } from './ng2-dropdown-state';
export class DropdownStateService {
    constructor() {
        this.menuState = {
            isVisible: /** @type {?} */ (false),
            /**
             * @return {?}
             */
            toString() {
                return this.isVisible === true ? 'visible' : 'hidden';
            }
        };
        this.dropdownState = new Ng2DropdownState();
    }
}
DropdownStateService.decorators = [
    { type: Injectable },
];
function DropdownStateService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DropdownStateService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DropdownStateService.ctorParameters;
    /** @type {?} */
    DropdownStateService.prototype.menuState;
    /** @type {?} */
    DropdownStateService.prototype.dropdownState;
}
