// ng2-material-dropdown/esm2015/src/modules/services/ng2-dropdown-state.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { EventEmitter } from '@angular/core';
export class Ng2DropdownState {
    constructor() {
        this.onItemSelected = new EventEmitter();
        this.onItemClicked = new EventEmitter();
        this.onItemDestroyed = new EventEmitter();
    }
    /**
     * \@name selectedItem
     * @desc getter for _selectedItem
     * @return {?}
     */
    get selectedItem() {
        return this._selectedItem;
    }
    /**
     * \@name selects a menu item and emits event
     * @param {?} item
     * @param {?=} dispatchEvent
     * @return {?}
     */
    select(item, dispatchEvent = true) {
        this._selectedItem = item;
        if (!dispatchEvent || !item) {
            return;
        }
        item.focus();
        this.onItemSelected.emit(item);
    }
    /**
     * \@name unselect
     * @desc sets _selectedItem as undefined
     * @return {?}
     */
    unselect() {
        this._selectedItem = undefined;
    }
}
function Ng2DropdownState_tsickle_Closure_declarations() {
    /** @type {?} */
    Ng2DropdownState.prototype.onItemSelected;
    /** @type {?} */
    Ng2DropdownState.prototype.onItemClicked;
    /** @type {?} */
    Ng2DropdownState.prototype.onItemDestroyed;
    /** @type {?} */
    Ng2DropdownState.prototype._selectedItem;
}
